### ISSUE TITLE
Homebrew should not install node if nvm is already installed

### ISSUE THREAD

**Issue by sjmueller** [yarnpkg/yarn#1116]
Attempted `brew install yarn` and it wanted to install node as a dependency. I already use Homebrew for nvm to manage my npm installs. Since nvm is already installed via Homebrew, the expected behavior is that nodejs installation should be skipped.

**Comment by billdybas** [comment]
You should make an issue on the [Homebrew repo](https://github.com/Homebrew/homebrew-core) or fix the [yarn formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/yarn.rb#L22).

**Comment by Daniel15** [comment]
I wonder if the Homebrew formula can check whether either node **or** nvm are installed?

**Comment by gillesdemey** [comment]
I would like to add [tj/n](https://github.com/tj/n) to the list, although we should just check if the `node` binary is installed.

**Comment by designorant** [comment]
I agree with @billdybas that this really belongs to Homebrew, but I want to chip in anyway as there's a bit more to it.

I think the current scenario is fine, and we shouldn't make it unnecessarily complex – every node version manager is different:
- What if there are no node versions installed via a manager?
- What if there are installed but none is set as global?
- What if there are two globally installed versions in two different places?

I've seen dozens of issues where [people ran into problems not knowing where their global packages are installed](https://github.com/infinitered/ignite/issues/382#issuecomment-248250114).

I've worked for quite a few companies with very different stacks over the years, and the most bulletproof setup for node I've found is to keep one always up to date version of node for globally installed packages (as we usually want to keep them up to date), and a manager for per-project ones.
 
I use [nodenv](https://github.com/nodenv/nodenv) (think of rbenv for node, it's pretty much 1:1 code) which I believe isn't as invasive as NVM nor as buggy as n, and flexible enough to let me specify a system node (handled by brew) along with managed per project versions if specified. That way all globally installed packages land in the same place, and can be updated without any issues. It also means that homebrew formulas depending on node work as expected, without adding another node version somewhere else.

**Comment by Daniel15** [comment]
>  the most bulletproof setup for node I've found is to keep one always up to date version of node for globally installed packages (as we usually want to keep them up to date), and a manager for per-project ones.

This sounds reasonable to me!

**Comment by sjmueller** [comment]
@designorant I like your philosophy, but let's not get too far from the core issue here, because it's pretty straightforward and simple. If you use homebrew to manage your node env (which many of us do), the two most popular options are node and nvm. Since this is the majority use case, people shouldn't be discouraged to `brew install yarn` if they went the nvm route because the formula doesn't count for this case.

**Comment by Daniel15** [comment]
I feel like this issue should go to the Homebrew repo rather than this repo. The solution should not be specific to Yarn, as it'd affect **any** app installed via Homebrew that requires Node.js. Ideally Homebrew itself should have some way of allowing apps to specify that they need a "node.js-ish environment", and have both Node.js and nvm fulfill that requirement.

Given that, I'm going to close this issue and kindly suggest that you ask the Homebrew maintainers about it. However, if you disagree with me, please let me know and I'd be more than happy to reopen this issue. Thank you for trying Yarn! 😄

**Comment by Daniel15** [comment]
Not sure if I reported it in the right place, but nonetheless I filed this issue: https://github.com/Homebrew/homebrew-core/issues/6458

**Comment by torifat** [comment]
I'm working on this: https://github.com/Homebrew/brew/pull/1408

**Comment by cnyyk** [comment]
agree with @gillesdemey.
If there's a node binary installed and version check is ok, `brew install yarn` should not install the node with homebrew again.

**Comment by Daniel15** [comment]
@cnyyk - @torifat  is working on this, as per his comment above 😄

**Comment by cnyyk** [comment]
Thanks for @Daniel15  and @torifat on this issue.
It seems that the changes made by @torifat are not been merged into the main branch of homebrew repo. 
I did the installation this morning but still show that dependency not installed.
Shown as below:
`
➜  ~ brew info yarn                         
yarn: stable 0.16.1 (bottled), HEAD
Javascript package manager
https://yarnpkg.com/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/yarn.rb
==> Dependencies
Required: node ✘

➜  ~ brew install yarn
==> Installing dependencies for yarn: node
==> Installing yarn dependency: node
==> Downloading https://homebrew.bintray.com/bottles/node-7.0.0.sierra.bottle.tar.gz
^C
➜  ~ brew install yarn --ignore-dependencies                                                                        
==> Downloading https://homebrew.bintray.com/bottles/yarn-0.16.1.sierra.bottle.tar.gz
Already downloaded: /Users/YCH/Library/Caches/Homebrew/yarn-0.16.1.sierra.bottle.tar.gz
==> Pouring yarn-0.16.1.sierra.bottle.tar.gz
🍺  /usr/local/Cellar/yarn/0.16.1: 4,257 files, 11.3M
➜  ~ 
`

**Comment by Daniel15** [comment]
Yeah, not merged yet 😄  You can subscribe to https://github.com/Homebrew/brew/pull/1408 if you like, that way you'll know when it's merged.

**Comment by cnyyk** [comment]
@Daniel15 Got it. Thanks.

**Comment by ljharb** [comment]
@Daniel15 this issue also deals with `nvm`, not just homebrew - how would you recommend `nvm` users avoid installing an extra `node`?

**Comment by Daniel15** [comment]
https://github.com/Homebrew/brew/pull/1408 should cover that I think.

**Comment by Daniel15** [comment]
How are you installing Yarn?
 - If you're installing through the Windows installer, it doesn't install Node.js
 - If you're installing through the Debian or Ubuntu packages, the Node.js package is marked as "recommended" so you can omit it by passing the `--no-install-recommends` flag to `apt-get install`
 - If you're using Homebrew, https://github.com/Homebrew/brew/pull/1408 should cover that

**Comment by ljharb** [comment]
Thanks :-) personally I prefer to install it with npm, but I appreciate you gathering all the advice into one place <3

**Comment by Daniel15** [comment]
Installing via npm should work too - That should just use whatever version of Node you're already using.

Note that installing via npm sometimes has strange edge cases, it's not signed via GPG, and it's significantly slower than any other installation method (that's not too big of a deal if you don't do it often. but I wouldn't do it in a build script for example).

**Comment by skalfyfan** [comment]
@Daniel15 Is there no way to *currently* to install yarn via Homebrew w/o installing Node?
I took a look at [Homebrew/brew#1408](https://github.com/Homebrew/brew/pull/1408) and looks like activity on getting a fix merged has stagnated?

**Comment by Daniel15** [comment]
Not sure, sorry. I don't use a Mac any more so I can't test it myself.

**Comment by Noiwex** [comment]
I don't need nodejs to be installed, if I needed this, I would install it from .pkg

**Comment by devotis** [comment]
Developer uses create-react-app. Developer sees yarn. Developer goes wtf is yarn. Developer goes  Google yarn. Developer sees `brew install yarn`. Developer runs `brew install yarn`. Developer drinks sip of coffee. Developer sees on screen and finds out node just called installed. Developer thinks wtf, I already had that with nvm. Developer freaks out having multiple node instances on computer. Developer Googles `brew install yarn without node`. Developer sees this issue and my post. Developer +1's my comment.

**Comment by sjmueller** [comment]
Exactly @devotis, x1000. However the folks at homebrew want to sit in their ivory tower on this one and shoot arrows at anyone who raises an alternate perspective: https://github.com/Homebrew/brew/pull/1408#issuecomment-272343833

**Comment by SenHeng** [comment]
@devotis 

it was mentioned [above](https://github.com/yarnpkg/yarn/issues/1116#issuecomment-258053685) by @cnyyk but you can do `brew install yarn --ignore-dependencies` though I don't think its necessary anymore. I've recently set up two new machines and managed to `brew install yarn` after installing nvm and haven't had any problems yet.

**Comment by Daniel15** [comment]
There's now a new flag (like `--no-node` or `--without-node` or something like that) for this reason. It's similar to `--ignore-dependencies` except it will remember it so when you upgrade Yarn it'll also avoid installing Node.js (rather than having to do `--ignore-dependencies` every time)

**Comment by atulsrathore** [comment]
For future Google searchers, use following to resolve the issue:

`brew install yarn --without-node`

Use `brew doctor` to verify there is no node warning.
NVM now manages your node versions.

**Comment by rojobuffalo** [comment]
For future Google searchers using a Brewfile.

`brew 'yarn', args: ['without-node']`

**Comment by ajoslin103** [comment]
> Developer uses create-react-app.
> Developer sees yarn.
> Developer goes wtf is yarn.
> Developer does Google: yarn.
> Developer sees `brew install yarn`.
> Developer runs `brew install yarn`.
> Developer drinks sip of coffee.
> Developer sees on screen and finds out Node.js just got installed.
> Developer thinks wtf (again), I already had that with nvm!
> Developer freaks out having multiple node instances on computer.
> Developer does Google: brew install yarn without node.
> Developer sees this issue only after the fact (like me).
> Developer +1's this comment.
> Developer goes wondering how to remove that redundant Node.js instance on computer.

Developer +1's this comment
Developer proceeds to remove Node 11 and replace with Node 10
Developer curses ivory arrows for not preferring LTS versions of Node by default
Developer installs nvm this time in the futile hope he'll not be back +1'ing this comment

**Comment by Noiwex** [comment]
Freed myself from homebrew repression, now I'm developing on windows, lmao.

**Comment by jaygorrell** [comment]
Was there a regression or change here? The official yarn docs call out the above flag, but running `brew install yarn --without-node` on a brand new environment is installing `node` as a dependency.

**Comment by Daniel15** [comment]
@jaygorrell  The Homebrew team changed that, and they appear to be locking any conversations about it, disallowing people from even talking about it:
 - https://github.com/Homebrew/homebrew-core/commit/a34c721586fece6d36d9b60fec5760e1b3361f11
 - https://github.com/Homebrew/homebrew-core/issues/31510
 - https://discourse.brew.sh/t/maintaining-a-tap-to-provide-options-for-a-formula/3871/28
😕

**Comment by Daniel15** [comment]
There's a discussion about it in https://github.com/yarnpkg/yarn/issues/6946. It sounds like we can create our own "tap" to re-add the `--without-node` option.

**Comment by jaygorrell** [comment]
Ah, thank you!

**Comment by Daniel15** [comment]
Yeah it's pretty annoying... Yarn is the fifth most popular package on Homebrew (ref https://formulae.brew.sh/analytics/install-on-request/365d/) yet they didn't inform us of this change at all.

**Comment by christiaanwesterbeek** [comment]
I just ran `brew upgrade yarn` and what I then saw made me almost write my first swear word ever on Github or any other online platform for that matter. Let's resort to all caps.

WHO DECIDED TO LET BREW INSTALL NODE ON DEVELOPER'S MACHINES WHEN THEY ALREADY HAVE NVM.

This sucks!

**Comment by SenHeng** [comment]
Considering the number of people who need to install yarn without node, it might be time to reconsider recommending brew as the official method. Maybe the [alternative] method should be the One True Way.

[alternative]: https://yarnpkg.com/en/docs/install#alternatives-stable

**Comment by Seebiscuit** [comment]
```[bash]
npm install --global yarn
```
FTW 💪

**Comment by Noiwex** [comment]
Homebrew just sucks

**Comment by 0xdevalias** [comment]
From what I could gather in all of the related threads/topics, the homebrew preference is to correctly set your `PATH` to use the binaries you want, which makes sense.

With that in mind, I've opened an issue to suggest that `brew`'s `depends_on` will similarly look at binaries available on `PATH` when considering whether it needs to install a new formulae.

This would seem to solve everyones needs in a pretty effective/non-hacky way IMO.

**Comment by Noiwex** [comment]
Guys, just stop using crapple products.

**Comment by sanjarcode** [comment]
I agree @billyjbryant. 

`nvm alias default someVersion` (set global default) breaks for me due to brew `node`, a dependency for MongoDB tools.
Quite funny considering the JS web app stack.

**Comment by Daniel15** [comment]
> > Considering the number of people who need to install yarn without node, it might be time to reconsider recommending brew as the official method. Maybe the [alternative](https://yarnpkg.com/en/docs/install#alternatives-stable) method should be the One True Way.
> 
> The issue isn't just with `yarn` any package that has `node` as a dependency should honor the path and not install node using brew if it already exists in path. This is not the case today and it breaks my setup anytime brew upgrades or installs anything that depends on node.


@billyjbryant  You should post in Homebrew's repo about this, not the Yarn repo.
